File: python/x402/src/x402_tron/signers/adapter.py
"""
Adapter that bridges agent-wallet providers to the KeyProvider protocol.
"""

from __future__ import annotations

from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from wallet import BaseProvider


class TronProviderAdapter:
    """Adapt agent-wallet's ``TronProvider`` (or ``FlashProvider``) to ``KeyProvider``.

    Reads ``address`` and ``_private_key_hex`` from the provider instance.

    Normally you don't use this directly — ``AgentWalletClientSigner`` and
    ``AgentWalletFacilitatorSigner`` wrap it automatically::

        from wallet import TronProvider
        from x402_tron.signers.client import AgentWalletClientSigner

        provider = TronProvider(private_key="deadbeef...")
        signer = AgentWalletClientSigner(provider, network="tron:nile")
    """

    def __init__(self, provider: BaseProvider) -> None:
        self._provider = provider

    def get_address(self) -> str:
        addr = getattr(self._provider, "address", None)
        if not addr:
            raise ValueError("Provider has no address (private key not set?)")
        return addr

    def get_private_key(self) -> str:
        pk = getattr(self._provider, "_private_key_hex", None)
        if not pk:
            raise ValueError("Provider has no private key")
        return pk[2:] if pk.startswith("0x") else pk

File: typescript/packages/x402/src/signers/adapter.ts
/**
 * Adapter that bridges agent-wallet providers to the KeyProvider interface.
 */

import type { KeyProvider } from './keyProvider.js';

/**
 * Adapt agent-wallet's TronProvider (or FlashProvider) to KeyProvider.
 *
 * Reads `address` and `tronWeb.defaultPrivateKey` from the provider.
 *
 * Normally you don't use this directly — `AgentWalletClientSigner`
 * wraps it automatically:
 *
 * @example
 * ```ts
 * import { TronProvider } from 'agent-wallet/wallet';
 * import { AgentWalletClientSigner } from '@bankofai/x402-tron';
 *
 * const provider = new TronProvider('https://nile.trongrid.io', ..., privateKey);
 * const signer = new AgentWalletClientSigner(tronWeb, provider, 'nile');
 * ```
 */
export class TronProviderAdapter implements KeyProvider {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  private provider: any;

  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  constructor(provider: any) {
    this.provider = provider;
  }

  getAddress(): string {
    const addr = this.provider.address;
    if (!addr) {
      throw new Error('Provider has no address (private key not set?)');
    }
    return addr as string;
  }

  getPrivateKey(): string {
    const pk: string | undefined = this.provider.tronWeb?.defaultPrivateKey;
    if (!pk) {
      throw new Error('Provider has no private key');
    }
    return pk.startsWith('0x') ? pk.slice(2) : pk;
  }
}
